metaflow/plugins/datatools/s3/s3op.py

```diff
@@ -322,22 +328,60 @@ def op_info(url):
                             # We indicate that the file was uploaded
                             result_file.write("%d %d\n" % (idx, 0))
                         except client_error as err:
-                            error_code = normalize_client_error(err)
-                            if error_code == 403:
-                                result_file.write(
-                                    "%d %d\n" % (idx, -ERROR_URL_ACCESS_DENIED)
-                                )
-                                continue
-                            elif error_code == 503:
-                                result_file.write("%d %d\n" % (idx, -ERROR_TRANSIENT))
-                                continue
-                            else:
-                                raise
+                            # shouldn't get here, but just in case
+                            handle_client_error(err, idx, result_file)
+                            continue
+                        except S3UploadFailedError as e:
+                            err = convert_to_client_error(e)
+                            handle_client_error(err, idx, result_file)
+                            continue
+                        except (SSLError, Exception) as e:
+                            # assume anything else is transient
+                            result_file.write("%d %d\n" % (idx, -ERROR_TRANSIENT))
+                            result_file.flush()
+                            continue
         except:
             traceback.print_exc()
+            result_file.flush()
             sys.exit(ERROR_WORKER_EXCEPTION)
 
 
+def convert_to_client_error(e):
+    match = re.search(
+        r"An error occurred \((\w+)\) when calling the (\w+) operation.*: (.+)", str(e)
+    )
+    if not match:
+        raise e
+    error_code = match.group(1)
+    operation_name = match.group(2)
+    error_message = match.group(3)
+    response = {
+        "Error": {
+            "Code": error_code,
+            "Message": error_message,
+        }
+    }
+    return ClientError(response, operation_name)
+
+
+def handle_client_error(err, idx, result_file):
+    error_code = normalize_client_error(err)
+    if error_code == 404:
+        result_file.write("%d %d\n" % (idx, -ERROR_URL_NOT_FOUND))
+        result_file.flush()
+    elif error_code == 403:
+        result_file.write("%d %d\n" % (idx, -ERROR_URL_ACCESS_DENIED))
+        result_file.flush()
+    elif error_code == 503:
```

Comment:
nit, could we merge the two cases here? Am fine also keeping it separate as a documentation type of thing.

Reply:
prefer to keep it this way (including some of the other branches that can be collapsed) for clarity